allennlp/data/tokenizers/huggingface_transformers_tokenizer.py

```diff
@@ -0,0 +1,430 @@
+
+from transformers.tokenization_auto import AutoTokenizer
+
+logger = logging.getLogger(__name__)
+
+CONTROL_CHARS = u"".join(
+    chr(c) for c in range(sys.maxunicode + 1) if unicodedata.category(chr(c))[0] == "C"
+)
+WHITESPACES = re.findall(r"\s", u"".join(chr(c) for c in range(sys.maxunicode + 1)), re.UNICODE)
+SPIECE_UNDERLINE = u"▁"
+
+
+@Tokenizer.register("huggingface_transformers")
+class HuggingfaceTransformersTokenizer(Tokenizer):
+    """
+    A ``HuggingfaceTransformersTokenizer`` uses a model from HuggingFace's
+    ``transformers`` library to tokenize some input text.  This often means wordpieces
+    (where ``'AllenNLP is awesome'`` might get split into ``['Allen', '##NL', '##P', 'is',
+    'awesome']``), but it could also use byte-pair encoding, or some other tokenization, depending
+    on the pretrained model that you're using.
+
+    We take a model name as an input parameter, which we will pass to
+    ``AutoTokenizer.from_pretrained``.
+
+    A suggestion for a usage *without* TokenIndexer, in order to fully utilize the Hugging Face's API:
+    1. Tokenize with ``tokenize`` or ``tokenize_with_offsets``
+    2. Encode the tokens with ``encode_plus``, in order to get the appropriate form the pretrained model expects
+    3. Create a LabelsField field with the encoded tokens
+    4. Optioanlly, add fields such as `token_type_ids` and `special_tokens_mask`
+
+    Example:
+        tokens = tokenizer.tokenize(text)
+        encoded_inputs = self._tokenizer.encode_plus([token.text for token in tokens],
+                                    add_special_tokens=True,
+                                    return_token_type_ids=True,
+                                    return_special_tokens_mask=True)
+        fields: Dict[str, Field] = {}
+        fields['tokens'] = LabelsField(encoded_inputs['input_ids'])
+        fields['token_type_ids'] = LabelsField(encoded_inputs['token_type_ids'])
+        fields['special_tokens_mask'] = LabelsField(encoded_inputs['special_tokens_mask'])
+        fields['pad_mask'] = LabelsField([1] * len(question_passage_tokens)) # For right padding
+
+    pretrained_model : ``str``
+        The name of the tokenizer to use.
+    init_kwargs : ``Dict[str, Any]
+        Arguments to pass to the constructor of the tokenizer
+    kwargs : ``Dict[str, Any]
+        Arguments to pass to the ``tokenize`` method
+    """
+
+    def __init__(
+        self, pretrained_model: str, init_kwargs: Dict[str, Any] = {}, kwargs: Dict[str, Any] = {}
+    ):
+        self._pretrained_model = pretrained_model
+        self._kwargs = kwargs
+        self._tokenizer = AutoTokenizer.from_pretrained(pretrained_model, **init_kwargs)
+        self._tokenization_cache: Dict[str, str] = {}
+
+    @overrides
+    def tokenize(self, text: str) -> List[Token]:
+        return [Token(text=token) for token in self._tokenizer.tokenize(text, **self._kwargs)]
+
+    def tokenize_with_offsets(self, text: str) -> List[Token]:
```

Comment:
Having this functionality would be great, but if it's going to go upstream, we should just wait for it to be merged there, instead of duplicating the code here.